### PR TITLE
Add LoongArch SIMD intrinsics

### PR DESCRIPTION
**Pull request by heiher**
This PR adds LoongArch SIMD (LSX/LASX) intrinsics and testcases.

### CODE REVIEW

**Review by Amanieu — COMMENTED** [review]
Are the tests also automatically generated, or just the intrinsic definitions?

File: crates/stdarch-gen/src/main.rs [hunk 1]
```diff
@@ -3183,6 +3186,10 @@ fn get_call(
 fn main() -> io::Result<()> {
     let args: Vec<String> = env::args().collect();
     let in_file = args.get(1).cloned().unwrap_or_else(|| IN.to_string());
+    let arch_out = args.get(2).cloned().unwrap_or_else(|| OUT.to_string());
+    if arch_out == "loongarch64" {
```
[the remark this answers is not in the document]
Comment (heiher) [reply crates/stdarch-gen/src/main.rs:3190]: Okay. I'll named the generator `stdarch-gen-${cpuarch}`:

* stdarch-gen-arm (neon)
* stdarch-gen-loongarch (lsx/lasx)

Thanks.

**Comment by heiher** [comment]
> Are the tests also automatically generated, or just the intrinsic definitions?

The tests are also automatically generated, but not by this generator.

**Comment by Amanieu** [comment]
Can you also include the generator used for making the tests? I don't want to include any auto-generated code in this repo without also including the generator used to create it.

Finally, can you also commit a copy of the header files used to generate the .spec files?

**Comment by heiher** [comment]
> Can you also include the generator used for making the tests? I don't want to include any auto-generated code in this repo without also including the generator used to create it.
> 
> Finally, can you also commit a copy of the header files used to generate the .spec files?

Thanks for your review. I have updated the patch to include code for generating test cases and also added copying of header files.

File: crates/core_arch/src/loongarch64/lasx/mod.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+#[rustfmt::skip]
+mod generated;
+
+#[rustfmt::skip]
+#[unstable(feature = "stdarch_loongarch", issue = "117427")]
+pub use self::generated::*;
+
+#[rustfmt::skip]
+mod tests;
```
Comment (Amanieu) [inline crates/core_arch/src/loongarch64/lasx/mod.rs:16]: Move the `#[cfg(test)]` here so it doesn't need to be in the generated code.

File: crates/stdarch-gen-loongarch/README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+LSX:
+```
+# Generate bindings
+OUT_DIR=`pwd`/crates/stdarch-gen-loongarch cargo run -p stdarch-gen-loongarch -- crates/stdarch-gen-loongarch/lsxintrin.h
+OUT_DIR=`pwd`/crates/core_arch cargo run -p stdarch-gen-loongarch -- crates/stdarch-gen-loongarch/lsx.spec
+
+# Generate tests
+OUT_DIR=`pwd`/crates/stdarch-gen-loongarch cargo run -p stdarch-gen-loongarch -- crates/stdarch-gen-loongarch/lsx.spec test
+# On LoongArch
+gcc -o lsx crates/stdarch-gen-loongarch/lsx.c -mlasx
```
Comment (Amanieu) [inline crates/stdarch-gen-loongarch/README.md:18]: Can you change these instructions to use qemu-loongarch64 and a cross-compiler (loongarch64-linux-gnu-gcc)? This makes it easier for people to run without access to loongarch hardware.

**Review by Amanieu — APPROVED** [review]
This looks great! Just a few minor changes and this is good to merge.

File: crates/stdarch-gen-loongarch/README.md (hunk 3 above)
Comment (heiher) [reply crates/stdarch-gen-loongarch/README.md]: I tested on [this git commit](https://github.com/qemu/qemu/commit/dccbaf0cc0f1744ffd7562a3dc60e4fc99fd9d44) of QEMU using qemu-loongarch64 and a cross-compiler (loongarch64-unknown-linux-gnu-gcc), and I was able to generate the correct test cases. Thanks.

File: crates/stdarch-gen-loongarch/README.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+LSX:
+```
+# Generate bindings
+OUT_DIR=`pwd`/crates/stdarch-gen-loongarch cargo run -p stdarch-gen-loongarch -- crates/stdarch-gen-loongarch/lsxintrin.h
+OUT_DIR=`pwd`/crates/core_arch cargo run -p stdarch-gen-loongarch -- crates/stdarch-gen-loongarch/lsx.spec
+
+# Generate tests
+OUT_DIR=`pwd`/crates/stdarch-gen-loongarch cargo run -p stdarch-gen-loongarch -- crates/stdarch-gen-loongarch/lsx.spec test
+# On LoongArch
```
Comment (heiher) [inline crates/stdarch-gen-loongarch/README.md:17]: I'll remove this line.

File: crates/core_arch/src/loongarch64/lasx/tests.rs [hunk 5]
```diff
@@ -0,0 +1,14692 @@
+// This code is automatically generated. DO NOT MODIFY.
+// See crates/stdarch-gen-loongarch/README.md
+
+mod tests {
```
Comment (Amanieu) [inline crates/core_arch/src/loongarch64/lasx/tests.rs:4]: This module is not needed, it is already inside a module called `tests`.

Comment (heiher) [reply crates/core_arch/src/loongarch64/lasx/tests.rs]: Done.

**Pull request merged**